sklearn/neighbors/classification.py

```diff
@@ -374,17 +391,51 @@ def predict(self, X):
             else:
                 mode = np.array([weighted_mode(pl, w)[0]
                                  for (pl, w)
-                                 in zip(pred_labels[inliers], weights[inliers])],
+                                 in zip(pred_labels[inliers],
+                                        weights[inliers])],
                                 dtype=np.int)
 
-            mode = mode.ravel()
+            # y_pred_sparse_multilabel = np.empty()
+            for k, classes_k in enumerate(classes_):
+                pred_labels = np.array([_y[ind, k].toarray()
+                                        for ind in neigh_ind[inliers]],
+                                       dtype=object)
+                y_pred_k = np.zeros(n_samples, dtype=np.int)
+                mode = self._mode(pred_labels, weights, inliers)
+                y_pred_k[inliers] = mode
 
-            y_pred[inliers, k] = classes_k.take(mode)
+                if outliers and self.outlier_label != 0:
+                    y_pred_k[outliers] = self.outlier_label
 
-        if outliers:
-            y_pred[outliers, :] = self.outlier_label
+                y_pred_sparse_multilabel.append(csc_matrix(y_pred_k).T)
 
-        if not self.outputs_2d_:
-            y_pred = y_pred.ravel()
+            y_pred = sparse.hstack(y_pred_sparse_multilabel)
+
+        else:
```

Comment:
I approved without reading properly. You have this else clause attaching to a for loop. Which is valid syntax, but not doing what you want it to do. I think that for loop is too long and nested to read, and it would be worth taking code out into a helper function.